### DISCUSSION TITLE
Integrating an RF modem interface with fprime-gds

### DISCUSSION THREAD

**Discussion by ReggieMarr** [nasa/fprime#2639] [General]
I'm currently working on integrating a previously developed interface to a Newtec MDM6000 modem which serves as the hardware interface between the flight software and the ground station (via RF).

I'd like to leverage the test integration api for deframing/decoding data received from the modem.

Since I'm interfacing with the modem via a udp socket connection it seems to me that I essentially want to provide a wrapper around my modem interface that can use the ThreadedUDPServer as a base class however I've been looking into the use of plugins lately and I wonder if instead I should be providing the modem interface via a plugin.

Any input on which seems to be the more "canonical" way is appreciated.

**Comment by Joshua-Anderson** [comment]
Is this interfacing with the modem on the FSW or the GDS side of things?

**Reply by ReggieMarr** [reply]
On the gds side of things. On the fsw side I have a transmitter component that takes telemetry, events, ECT and frames it before sending it to the lower level drivers for transmission as a byte stream

**Reply by Joshua-Anderson** [reply]
Yeah, plugin support for your modem communication adapter would be the ideal way to go so that you don't have to modify the GDS code itself. Communication adapters are supported by the plugin architecture, but it looks like there aren't any communication plugin examples yet. However, the built-in [IP](https://github.com/fprime-community/fprime-gds/blob/devel/src/fprime_gds/common/communication/adapters/ip.py) and UART adapters are packaged as plugins and could be a good reference

If you haven't found it yet, there's a plugin how-to guide [here](https://github.com/nasa/fprime/blob/devel/docs/HowTo/develop-gds-plugins.md)

I suspect you could probably even use the ThreadedUDPServer class as part of your plugin if that's helpful

**Reply by LeStarch** [reply]
@Joshua-Anderson is spot-on.  I definitely recommend using plugins. A few helpful tips as plugins are new functionality:

1. You'll want to clone https://github.com/fprime-community/fprime-gds.git and install it with `pip install /path/to/clone`.  
2. Use the "communication" plugin category to setup a custom reader of UDP
3. Use the "framing" plugin category if you need to frame/deframe non-standard frames (not using the fprime frame format)
4. Ask questions!  These plugins are for your purpose....but you are an initial user.  We are here to help!

**Reply by LeStarch** [reply]
Oh, and any feedback is appreciated!

**Reply by ReggieMarr** [reply]
looks great ! I'll update here as I'm sure I'll have questions as I get going (If that seems appropriate).
I actually started off thinking I'd have to maintain a forked version of fprime-gds so this def seems like a better way to go

**Reply by ReggieMarr** [reply]
So I finally got back to this and I think I'm missing some info about how I can package this with fprime-gds (I see some mention in the docs but it's a bit unclear to me)

I have a modemadapter like this:
```python
#!/usr/bin/env python3
import abc
import socket
from typing import NamedTuple, Tuple, Type
import threading
from fprime_gds.common.communication.framing import FramerDeframer
from fprime_gds.plugin.definitions import gds_plugin_implementation
from fprime_gds.common.communication.adapters import base as fprime_comms_base
from fprime_gds.common.communication.adapters.ip import IpAdapter
from modem_interface import ModemInterface
from dataclasses import dataclass, field
from datetime import datetime
from time import time

@dataclass
class ModemAdapterCfg:
    """
    A class for the emulator's status.
    """

    modem_ip_address: Tuple[str, int]
    fsw_ip_address: Tuple[str, int]
    poll_hz: int = 10

class ModemAdapter(fprime_comms_base):
    '''This provides an fprime-gds adapter plugin such that downlink is facilitated via rf
    (through the use of the modem interface) and uplink is meant to be done using the usual tcp interface.
    '''

    def __init__(self, cfg: ModemAdapterCfg):
        self.cfg = cfg
        self.poll = threading.Event()
        self.shutdown = threading.Event()
        self.modem = ModemInterface(cfg.modem_ip_address[0], cfg.modem_ip_address[1])
        self.ip = IpAdapter(cfg.fsw_ip_address[0], cfg.fsw_ip_address[1])

    def __poll_for_datagrams(self):
        while not self.shutdown.is_set()
            if self.poll.is_set():
                self.modem.decodeIncomingData()
            time.sleep(1 / self.cfg.poll_hz)

    def __del__(self):
        self.close()
        self.shutdown.set()

    def open(self):
        self.poll.set()

    def close(self):
        self.poll.clear()

    @abc.abstractmethod
    def read(self, timeout=0.500):
        return self.modem.popDatagram()

    @abc.abstractmethod
    def write(self, frame):
        """
        Write to the interface. Must be overridden by the child adapter. Throw no fatal errors, reconnect instead.

        :param frame: framed data to uplink
        :return: True if data sent through adapter, False otherwise
        """
        self.ip.write(frame)

    @classmethod
    @gds_plugin_specification
    def register_communication_plugin(cls) -> Type["ModemAdapter"]:
        return cls

```

Where I want to use rf for the downlink and ip for the uplink (ideally I'd actually be able to get downlink via ip simulataneously for comparison but that for a later step)

Currently I'm attempting to test the tlm with just Ip with the following:
```python
class TelemLink(object):
    """
    Class for managing telemetry link.
    """

    def __init__(self, cfg: TelemLinkCfg, logger: Logger) -> None:
        """
        Initializes a TelemLink object with the provided configuration
        and logger objects.
        """
        self.fprime_api, self.fprime_pipeline = self._establish_connection(cfg, logger)

        self.cfg = cfg
        self.logger = logger

    def __del__(self):
        self.fprime_api.teardown()
        self.fprime_pipeline.disconnect()

    @staticmethod
    def _establish_connection(cfg: TelemLinkCfg, logger: Logger) -> Tuple[IntegrationTestAPI, StandardPipeline]:
        """
        Establishes connection with the api interface by setting up
        the pipeline using the provided configuration and logger.
        """
        # Instantiate a pipeline and config manager for managing the api interface
        pipeline = StandardPipeline()
        config_manager = ConfigManager()
        # Launch tcp server
        run_deployment.launch_tts(50000, "0.0.0.0", cfg.tmp_path)

        # Setup pipeline
        pipeline.setup(config_manager, cfg.dict_path, cfg.tmp_path)

        api_connect_str = f"tcp://127.0.0.1:50000"
        api = IntegrationTestAPI(pipeline, cfg.log_path)
        logger.info(f"Attempting to connect to {api_connect_str}")
        pipeline.connect(api_connect_str, RoutingTag.FSW)

        return api, pipeline

    def get_data_mock_stats(self) -> DataMockStats:
        """
        Acquires telemetry from the link, logs the result and asserts
        conditions based on the received telemetry.
        Displays telemetry based on the preconfigured export type
        """
        result = self.fprime_api.await_telemetry("DataMockStatsTlm")
        self.logger.info(result)
        if result is None:
            return None
        else:
            return DataMockStats.from_dict(result.get_val())

    def set_flight_downlink_profiling_enabled(self, is_enabled: bool):
        """
        Sends the `SetIsFlightDownlinkProfilingEnabled` command to the Profiler component.
        """
        result = self.fprime_api.send_command("profiler.SetIsFlightDownlinkProfilingEnabled", [is_enabled])

        self.logger.info(result)
        return result

```

How can I pass through the modem config at this level and "select" it as the adapter of choice ?

**Comment by ReggieMarr** [comment]
I've expanded on my initial implementation a bit but I'll leave the reply comment up for posterity.
This is what I have now:
```python
#!/usr/bin/env python3
import abc
import socket
from typing import NamedTuple, Tuple, Type
import threading
from fprime_gds.common.communication.framing import FramerDeframer
from fprime_gds.common.communication.ground import TCPGround
from fprime_gds.common.communication.updown import Downlinker, Uplinker
from fprime_gds.plugin.definitions import gds_plugin_implementation
from fprime_gds.common.communication.adapters.base import BaseAdapter
from fprime_gds.common.communication.adapters.ip import IpAdapter
from fprime_gds.plugin.system import FpFramerDeframer
from modem_interface import ModemInterface
from dataclasses import dataclass, field
from datetime import datetime
from time import time
import sys

@dataclass
class ModemAdapterCfg:
    modem_ip_address: Tuple[str, int]
    fsw_ip_address: Tuple[str, int]
    poll_hz: int = 10
    poll_timeout_s = 5000

@dataclass
class ModemStatus:
    bytes_decoded: int = 0
    start: datetime = datetime.now()
    isPolling: bool = False

    def __str__(self):
        """
        Returns a formatted string representation of the status.
        """
        time_since_start = 0
        if self.isPolling:
            time_since_start = (datetime.now() - self.start).total_seconds()
        return (
            f"Status:\n"
            f"  Time since start (S): {time_since_start} ms\n"
            f"  Bytes Decoded: {self.bytes_decoded}\n"
        )

class ModemAdapter(BaseAdapter):
    '''This provides an fprime-gds adapter plugin such that downlink is facilitated via rf
    (through the use of the modem interface) and uplink is meant to be done using the usual tcp interface.
    '''

    def __init__(self, cfg: ModemAdapterCfg):
        self.cfg = cfg
        self.poll = threading.Event()
        self.shutdown = threading.Event()
        self.modem = ModemInterface(cfg.modem_ip_address[0], cfg.modem_ip_address[1])
        self.ip = IpAdapter(cfg.fsw_ip_address[0], cfg.fsw_ip_address[1])
        self.poll_thread = threading.Thread(target=self._poll_for_datagrams, daemon=True)
        self._status = ModemStatus()

    @classmethod
    def get_name(cls) -> str:
        return "ModemAdapter"

    @classmethod
    def get_arguments(cls):
        """ Arguments to request from the CLI """
        return {}

    @classmethod
    def check_arguments(cls):
        """ Check arguments from the CLI """
        pass

    @property
    def status(self):
        """
        Retrieves and logs the status of the modem adapter.
        """
        return self._status

    def _poll_for_datagrams(self):
        while not self.shutdown.is_set():
            # Blocks thread on max timeout
            try:
                self.poll.wait(self.cfg.poll_timeout_s)
                self.modem.decodeIncomingData()
                time.sleep(1 / self.cfg.poll_hz)
            except RuntimeError:
                pass

    def __del__(self):
        self.close()
        self.shutdown.set()

    def open(self):
        self._status.start = datetime.now()
        self.poll.set()

    def close(self):
        self.poll.clear()

    def read(self, timeout=0.500):
        """
        Function that checks for available datagrams and pops the first one
        when it becomes available. If no datagram becomes available within
        'timeout' seconds, a RuntimeError is raised indicating the timeout.

        Parameters:
        timeout (float): Maximum time in seconds to wait for a datagram.
            Default is 0.500 seconds.

        Raises:
        RuntimeError: When the function times out without a datagram being
            available.

        Returns:
        datagram: The first datagram that becomes available.
        """
        start_time = time.time()  # recording the start time

        datagram = None
        while time.time() - start_time < timeout:
            if self.modem.datagramsAvailable():
                datagram = self.modem.popDatagram()
                return datagram
            # Sleep for a small duration to avoid busy-waiting
            time.sleep(0.01)

        end_time = time.time()  # recording the end time

        raise RuntimeError(f"Modem timed out after {end_time - start_time} seconds!")

    def write(self, frame):
        self.ip.write(frame)

    @classmethod
    @gds_plugin_implementation
    def register_communication_plugin(cls) -> Type["ModemAdapter"]:
        return cls

def launch_modem(modem: ModemAdapter):
    adapter = modem

    ground = TCPGround()
    checksum_type='crc32',
    framer = FpFramerDeframer(checksum_type)
    downlinker = Downlinker(modem, ground, framer)
    uplinker = Uplinker(adapter, ground, framer, downlinker)

    # Open resources for the handlers on either side, this prepares the resources needed for reading/writing data
    ground.open()
    adapter.open()

    # Finally start the processing of uplink and downlink
    downlinker.start()
    uplinker.start()

    uplinker.join()
    downlinker.join()

```

Does this essentially capture the approach that the plugin framework implies ?

**Reply by ReggieMarr** [reply]
@LeStarch would you be able to comment on my implementation here? My approach has been to implement something similar to what's found in ```launch_comm``` but I don't know if that's correct

**Reply by LeStarch** [reply]
The plugin for this type of Comm component is launched inside of the `comm` process.  Is there a reason you cannot just run:

```
fprime-gds --comm-selection ModemAdapter
```

That should start up the standard GDS, but swap out the comm adapter for the modem you have above.

**Reply by LeStarch** [reply]
Another point: the plugin system applies arguments to the `__init__` method as defined and returned from the `get_arguments` method and checked with the `check_arguments` method. I'd break them out like this:

```python
class ModemAdapter():

def __init__(self,  modem_ip_address):
  ...
  self.modem = ModemInterface(modem_ip_address)
  

def check_arguments(cls, **modem_cfg):
    ....

def get_arguments():
   return { ("--modem-ip-address", ) : {
      "help": "Ip address of the modem",
      "type": str,
    }
  }

```
>  Note:
> 1. This is pseudocode
> 2. Shows both "keyword arguments" approach, and "explicit argument" approach, choose one or the other

Hope this gives some insight.  I am here for more questions.

**Reply by ReggieMarr** [reply]
> The plugin for this type of Comm component is launched inside of the `comm` process. Is there a reason you cannot just run:
> 
> ```
> fprime-gds --comm-selection ModemAdapter
> ```
> 
> That should start up the standard GDS, but swap out the comm adapter for the modem you have above.

So I understand what was needed for the args to work however I'm still not understanding how to build/pip install with my plugin so that this is available when I run ```fprime-gds```.

For example I have this ```pyproject.toml```

```toml
[build-system]
requires = ["setuptools>=61", "wheel"]
build-backend = "setuptools.build_meta"

[project]
name = "rdx-gds-plugin"
authors = [
  {name = "Reginald Marr"},
]

requires-python = ">= 3.8"
dynamic = ["dependencies", "version"]

[entry_points."fprime_gds"]
modem_adapter = "modem_adapter:ModemAdapter"

[tool.setuptools_scm]
```
with a setup.py that references the local copy of fprime-gds

however when I go to run fprime-gds after installing I get this:

```bash
❯ fprime-gds -n --comm-selection ModemAdapter          
usage: fprime-gds [-h] [-l LOGS] [--log-directly] [--log-to-stdout] [-g {none,html}] [--dictionary DICTIONARY]
                  [-c CONFIG] [--gui-addr GUI_ADDR] [--gui-port GUI_PORT] [--zmq] [--zmq-server]
                  [--zmq-transport serverInUrl serverOutUrl] [--tts-port TTS_PORT] [--tts-addr TTS_ADDR] [-n]
                  [--app APP] [-r ROOT_DIR] [--ip-address ADDRESS] [--ip-port PORT] [--comm-adapter {none,ip}]
                  [--comm-checksum-type {fixed,crc32}]
fprime-gds: error: unrecognized arguments: --comm-selection ModemAdapter
```

Which to me indicates that the plugin isn't being found

**Reply by LeStarch** [reply]
Have you installed it into your environment?

```
cd /directory/that/contains/pyproject/toml/file
pip install -e .
```
> The `-e` does a developer install, allowing you to change files without needing to reinstall.

**Reply by LeStarch** [reply]
Looks like I need to add a note to the how-to guide on that step...  Good catch!

**Reply by ReggieMarr** [reply]
I did do that, but for some reason my plugin doesn't show up in the list of available plugins, I wonder could you explain a bit about how this works ?

```python
class PluginArgumentParser(ParserBase):
    """Parser for arguments coming from plugins"""

    DESCRIPTION = "Plugin options"
    FPRIME_CHOICES = {
        "framing": "fprime",
        "communication": "ip",
    }

    def __init__(self):
        """Initialize the plugin information for this parser"""
        self._plugin_map = {
            category: Plugins.system().get_plugins(category)
            for category in Plugins.system().get_categories()
        }


```

I'm trying to wrap my head around how plugins are detected.

Note that as I'm trying to debug this I'm doing so via this:

```json
{
    "name": "Debug fprime-gds src",
    "type": "python",
    "request": "launch",
    "program": "${workspaceFolder}/gds/custom-deps/fprime-gds/src/fprime_gds/executables/run_deployment.py",
    "env": {
        "PYTHONPATH": "${env:PYTHONPATH}:${workspaceFolder}/gds/custom-deps/fprime-gds/src/"
    },
    "console": "integratedTerminal",
    "args": ["-n", "--ip-address", "172.30.1.2", "--ip-port", "50000",
             "--dictionary", "${workspaceFolder}/gds/Ref/FlightComputerTopologyAppDictionary.xml"],
    "justMyCode": false,
    "cwd": "${workspaceFolder}"
},
{
    "name": "Debug fprime-gds module",
    "type": "python",
    "request": "launch",
    "module": "fprime_gds.executables.run_deployment",
    "console": "integratedTerminal",
    "args": ["-n"],
    "justMyCode": false,
    "cwd": "${workspaceFolder}"
},

```

Which essentially allows me to run fprime-gds either as an installed module (installed via my plugin) or by calling run_deployment directly (which since I'm doing this with [dap-mode](https://github.com/emacs-lsp/dap-mode) it does so with the python interpreter associated with the pyenv that I installed my plugin/fprime-gds in)

**Reply by LeStarch** [reply]
Some things I've been doing on my end:

1. I did a PR to clean-up that guide more: https://github.com/nasa/fprime/pull/2670
2. I did a pre-release of fprime-gds `v3.4.4a1` such that you can pip install your plugins with clean dependencies
3. I published examples: https://github.com/fprime-community/fprime-gds-plugin-examples

Some things you can try:
1. `pip install fprime-gds==v3.4.4a1` to make sure you have the latest plugin support in your environment
2. Try installing the example above.  This will show if basic plugins are working in your environment.  They don't do much, but will prove the mechanisms before diving into more complex examples
3. Review the above PR.  There were some typos and mistakes that I've correct based on your testing.

**Reply by LeStarch** [reply]
Now to answer your question on how this all works.  We use Pluggy to load plugins.  The process looks like this:

1. Install a plugin via an entry-point of a python package
2. GDS launches and calls the plugin argument parser
3. Argument parser asks plugin system for valid plugins
  1. Plugin are validated to ensure that the plugin registration function returns a non-virtual subclass of the necessary virtual base class (i.e. a framing plugin returns a concrete class that "is-a" `FramerDeframer`)
4. For each active plugin:
  1. `get_name` is called to determine the name
  2. `get_arguments` is called to get arguments
  3.  A parser is built to handle arguments
5. During argument parser resolution, the plugin class is instantiated with CLI supplied arguments
6. During GDS execution where a plugin is supported (e.g. framing) the plugin instance is supplied to that sub-system

**Reply by ReggieMarr** [reply]
Awesome, took a brief look at the docs, I may add some comments if that's alright ? 

The example project is especially helpful, I realized my issue was in my pyproject.toml I had

```toml
[entry_points."fprime_gds"]
```

Where I should have had 
```toml
[project.entry_points.fprime_gds]
```

Now it seems to work as expected.

One comment I was going to make was that it seems like having the docs for fprime-gds plugins in the fprime repo, seperately from the fprime-gds repo itself and then the example code is found in yet another repo.

Is the idea to at some point to collect this into a super project ? Or what's the reasoning to keep all this stuff seperate

**Reply by LeStarch** [reply]
Currently, all our documentation is hosted out of fprime (hence the documentation being there) as it hosts our website.

It might be reasonable to collect examples within the `fprime-gds` repository itself.  I kept them as a separate repository but could easily make them a sub-folder under `fprime-gds`.

Would it make more sense to put the examples within that repository?  I've seen both example directories and example repositories so I don't have a strong opinion.

**Reply by SterlingPeet** [reply]
Personally, I think that using an Out-Of-Tree format example shows exactly how a 3rd party implementation _should be done_ and therefore carries value.  I don't want to be required to entangle my own [admittedly currently non-existent] GDS plugins into the upstream code, and that might be hard to sort out without an example in the expected format and layout.  The documentation should indicate that is the intent, and all of the involved repositories should link to the relevant documentation at or near the top of the repository README.

NOTE: I have not checked that documentation is currently good or poor, I am just pointing to the fact the _clarity is paramount_.

**Reply by ReggieMarr** [reply]
I think if it were me doing it I would have kept it in the same repo for discoverablitys sake (also since the examples could be tracked and updated as fprime-gds changes (especially useful for in the future if we were to have breaking changes).
I see the point that it being in a seperate repo is a truer representation of how a 3rd party would implement plugins but that does come with some costs.

I could see having the examples as a submodule being a decent compromise though.

The documentation could also leverage submodules I would think. Submodules do come with their own sorts of headaches admittedly but it is nice to know for sure when you open the repo that the documentation is representative of a particular version of the underlying libraries they represent.

**Reply by LeStarch** [reply]
I can say that the example repo links to the documentation.  This meets point 1.  We could easily add a sub-repository link for examples as that would solve both issues: discoverability, and clear-separation. @thomas-bc do you have an opinion on this?

**Reply by ReggieMarr** [reply]
another point I just wanted to add wrt to fprime-gds and fprime. Is there anything to track compatibility between the two (like producing errors in case fprime-gds detects that the flight software is running a backwards incompatible version of fprime or visa versa)

**Reply by SterlingPeet** [reply]
Tracking compatibility is difficult, yes.  @LeStarch this points towards my commentary about rolling a lightweight package manager to handle dependencies, otherwise the separation of repositories that supports Separation of Concerns does decrease the discoverability without tooling to help fill the gap.

The desire/need for ease of discoverability annoyingly pulls the opposite direction from having cleanly separated repositories to allow projects to mix in proprietary solutions without needing to pull in and own the framework parts too.

We don't put all of the python modules you need for a project in one repository so we can get project done, we have a package management tool to solve that (pip).  This tension represents a bit of nash equilibrium where incremental changes one direction or the other are hard on somebody.

**Reply by ReggieMarr** [reply]
Hmm, I see what you're saying. imho a package manager is a bit overkill here though. I can say in my own project I have a structure like this:

```
├── Dockerfile
├── downlink-client
│   ├── docker-compose.yml
│   ├── Dockerfile
│   ├── gds
│   │   ├── custom-deps
│   │   │   └── fprime-gds
│   ├── pyrightconfig.json
│   └── README.md
├── embedded-nodes-firmware
│   ├── common
│   ├── libs
│   ├── nodes
│   └── README.md
├── flightcomputer-software
│   ├── docker-compose.yml
│   ├── Dockerfile
│   ├── FlightComputer
│   ├── fprime
│   ├── pyrightconfig.json
│   └── README.md
├── README.md
└── tools
    └── generate_release_description.py
```
where downlink-client, flightcomputer-software, embedded-nodes-firmware is how the 3 main projects are stored (i.e. our ground system software, low level device and flight computer software respectively) Note that each of those are submodules/sub-repos.

Then within flightcomputer-software and downlink-client I have fprime and fprime-gds as submodules.

Since there's no release guide or lts I kinda find myself in a situation where I really don't want to change the fprime submodule to something other than the first thing that I got to work. However on the other side since there''s lot of plugin development happening I find myself updating the version of fprime-gds.

However since fprime-gds's documentation is in fprime you end up needing to checkout a branch of fprime that is newer than what you will use in the flightcomputer software just to read the documentation that contains references to features used in fprime-gds.

I get that we need to be able to make a relative reference to the fprime-gds docs from fprime for the purposes of generating the website however this could still be accomplished if fprime-gds is a submodule of fprime and it has its own docs folder. 

Wrt to the python example I would say that the manner in which I manage python dependencies depends a bit on how I'm using them. If it's something simple then I'd probably track the version in a pyproject.toml or something and then just pip install it.
However if its a dependency that it more wide ranging that I find myself reviewing the source code for or potentially making changes to it then I would track it as a submodule.

In some senses this actually points out that we already have package management in the form of cmakelists.txt and/or pyproject.toml.

Thats just my two cents though if you think the way I'm doing it is out of left field I'm happy to hear of a better approach.

**Reply by SterlingPeet** [reply]
@ReggieMarr I think your methodology is valid.  As someone who does devops as part of my job responsibilities, I prefer methods that emphasize tracking release versions as artifacts rather than source trees.  That is orthogonal to the need to review foreign (not in-house development) code.

Some differences that crop up here: when developing a solution, interoperability beyond the specific versions you develop against is irrelevant.  This means convenience, simplification, or consistency in process can take precedence with respect to dependency management.  For a library or framework developer (like @LeStarch ), the end user/developer setup is not constrained and adoption may depend on dependency version flexibility.  My point here is that the F Prime project solution to dependency management may not be the same as yours, @ReggieMarr, nor need it be.  Your method looks perfectly valid to me, especially in the context of a solution/mission developer.

**Reply by ReggieMarr** [reply]
I see, yeah my usecase is just one of many. I don't mean to convey that having fprime and fprime-gds as submodules in my "3rd party project" is a result tracking fprime-gds as a submodule of fprime.

My point is rather I don't see the need for a package manager here. Imho pip and cmake provide a convenient and familiar way of installing specified versions. When it comes to selecting the version we would ideally have a detailed changelog, release notes, and release schedule to inform what versions should interop with one another.

Given it would probably be non-trivial to produce those docs I was suggesting that tracking fprime-gds as a submodule to fprime would be an easy way of conveying at least some level of interoperability, this also wouldn't necessarily be a solution "in-tree" since by default cloning fprime wouldn't include the submodule, just a reference to it. And then we'd also have the benefits of discoverability when desired but not including irrelevant docs when not (i.e if someone doesn't care about fprime-gds they won't find references to it in fprime, but if they do they'll found in a manner consistent with fprime).

**Reply by SterlingPeet** [reply]
Yes, I think that an emerging issue in F Prime is that tracking version compatibility is challenging.  I do agree that your suggestion does solve it as it currently itches!

I have advocated for separating out all of the components in the framework with versioning information built around functionality and port definitions (so it can be automated to some degree) but thats not something that is currently on the F Prime roadmap to my knowledge.   Perhaps that helps you see where I am coming from, and that I am not implying that the extreme solution I suggested is an immediate need.

**Reply by ReggieMarr** [reply]
I see yeah I can see how that sort of separation would benefit from some automation and could help for more expansive projects.
